### PR TITLE
Clean up worker pools before closing any leftover quic sessions

### LINKED ISSUE #482: Crash in msquic!MsQuicSessionClose
### Describe the bug

Crash on shutdown of spinquic.
```
 # Child-SP          RetAddr           Call Site
...
0a 000000a3`3ecfe980 00007ffe`8b688fec ntdll!KiUserExceptionDispatcher+0x2e
0b 000000a3`3ecff750 00007ffe`8b67c084 msquic!MsQuicSessionClose+0x19c [D:\a\1\s\src\core\session.c @ 335] 
0c 000000a3`3ecff7b0 00007ffe`8b67c468 msquic!MsQuicLibraryUninitialize+0xa4 [D:\a\1\s\src\core\library.c @ 276] 
0d 000000a3`3ecff800 00007ffe`8b679575 msquic!MsQuicRelease+0xc8 [D:\a\1\s\src\core\library.c @ 389] 
0e 000000a3`3ecff840 00007ff7`9f575612 msquic!MsQuicClose+0xa5 [D:\a\1\s\src\core\library.c @ 1079] 
0f 000000a3`3ecffaa0 00007ff7`9f58c559 spinquic!main+0xb02 [D:\a\1\s\src\tools\spin\spinquic.cpp @ 829] 
10 000000a3`3ecffdf0 00007ff7`9f58c43e spinquic!invoke_main+0x39 [d:\A01\_work\6\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl @ 79] 
11 000000a3`3ecffe40 00007ff7`9f58c2fe spinquic!__scrt_common_main_seh+0x12e [d:\A01\_work\6\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl @ 288] 
12 000000a3`3ecffeb0 00007ff7`9f58c5e9 spinquic!__scrt_common_main+0xe [d:\A01\_work\6\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl @ 331] 
13 000000a3`3ecffee0 00007ffe`c45a7974 spinquic!mainCRTStartup+0x9 [d:\A01\_work\6\s\src\vctools\crt\vcstartup\src\startup\exe_main.cpp @ 17] 
14 000000a3`3ecfff10 00007ffe`c714a271 kernel32!BaseThreadInitThunk+0x14
15 000000a3`3ecfff40 00000000`00000000 ntdll!RtlUserThreadStart+0x21
```

### Detailed Info

It looks like there is unsynchronized access in this code (session.c):

```
    } else {
        //
        // This is the global unregistered session. All connections need to be
        // immediately cleaned up.
        //
        QUIC_LIST_ENTRY* Entry = Session->Connections.Flink;
        while (Entry != &Session->Connections) {
            QUIC_CONNECTION* Connection =
                QUIC_CONTAINING_RECORD(Entry, QUIC_CONNECTION, SessionLink);
            Entry = Entry->Flink;                                            <=== CRASH HERE
            QuicConnOnShutdownComplete(Connection);
        }
    }
```

My guess is that a datapath thread or worker thread modified this session's connection list in parallel. Need to take a closer look at the synchronization around this code path.

### PR DESCRIPTION
**Pull request by ThadHouse**
Without this, we race cleanup.

Should fix #482

### CODE REVIEW

File: src/core/session.c [hunk 1]
```diff
@@ -326,15 +326,10 @@ MsQuicSessionClose(
...
+        // on the worker queue.
         //
-        QUIC_LIST_ENTRY* Entry = Session->Connections.Flink;
-        while (Entry != &Session->Connections) {
-            QUIC_CONNECTION* Connection =
-                QUIC_CONTAINING_RECORD(Entry, QUIC_CONNECTION, SessionLink);
-            Entry = Entry->Flink;
-            QuicConnOnShutdownComplete(Connection);
-        }
+        MsQuicSessionShutdown(Handle, QUIC_CONNECTION_SHUTDOWN_FLAG_NONE, 0);
```
Comment (nibanks) [inline src/core/session.c:332]: You should be using the silent flag here.

**Comment by nibanks** [comment]
Also, as a part of this PR, we should update spinquic to go through the complete library shutdown path several times throughout the run. Otherwise, we aren't getting much test coverage of the area this is supposed to fix.

File: src/tools/spin/spinquic.cpp [hunk 2]
```diff
@@ -731,103 +733,108 @@ main(int argc, char **argv)
     TryGetValue(argc, argv, "seed", &RngSeed);
     srand(RngSeed);
 
-    for (size_t i = 0; i < BufferCount; ++i) {
-        Buffers[i].Length = MaxBufferSizes[i]; // TODO - Randomize?
-        Buffers[i].Buffer = (uint8_t*)malloc(Buffers[i].Length);
-        EXIT_ON_NOT(Buffers[i].Buffer);
-    }
-
     SpinQuicWatchdog Watchdog((uint32_t)Settings.RunTimeMs + WATCHDOG_WIGGLE_ROOM);
```
Comment (nibanks) [inline src/tools/spin/spinquic.cpp:736]: The threads are still configured to run `Settings.RunTimeMs` right? So if they do it more than once, the watchdog would fire, right?

Comment (ThadHouse) [reply src/tools/spin/spinquic.cpp]: Good point. Moved it into the loop.

Comment (nibanks) [reply src/tools/spin/spinquic.cpp]: I don't want the number of loops to multiple the total run time. I'd rather keep the watchdog where it is, and divide the time each loop runs.

**Pull request merged**